File: backends/dotnet/src/Shared/Web/Result.cs
namespace StackBraid.Shared.Web;

/// <summary>
/// A handler's outcome: exactly one of a value or an <see cref="AppError"/>.
/// Endpoints translate a failure into the contract's Problem envelope
/// (see <c>ProblemDetailsMapper</c>) and a success into the response schema
/// directly — the contract has no generic success wrapper, only DTOs.
/// </summary>
public readonly struct Result<T>
{
    public bool IsSuccess { get; }
    public T? Value { get; }
    public AppError? Error { get; }

    private Result(bool isSuccess, T? value, AppError? error)
    {
        IsSuccess = isSuccess;
        Value = value;
        Error = error;
    }

    public static Result<T> Success(T value) => new(true, value, null);
    public static Result<T> Failure(AppError error) => new(false, default, error);

    public static implicit operator Result<T>(AppError error) => Failure(error);

    public TResult Match<TResult>(Func<T, TResult> onSuccess, Func<AppError, TResult> onFailure) =>
        IsSuccess ? onSuccess(Value!) : onFailure(Error!);
}

/// <summary>Non-generic form for a use case that returns nothing on success.</summary>
public readonly struct Result
{
    public bool IsSuccess { get; }
    public AppError? Error { get; }

    private Result(bool isSuccess, AppError? error)
    {
        IsSuccess = isSuccess;
        Error = error;
    }

    public static Result Success() => new(true, null);
    public static Result Failure(AppError error) => new(false, error);

    public static implicit operator Result(AppError error) => Failure(error);

    public TResult Match<TResult>(Func<TResult> onSuccess, Func<AppError, TResult> onFailure) =>
        IsSuccess ? onSuccess() : onFailure(Error!);
}

File: backends/python/src/app/shared/web/result.py
"""A handler's outcome: exactly one of a value or an ``AppError``.

Endpoints translate a failure into the contract's Problem envelope (see
``problem.py``) and a success into the response schema directly — the
contract has no generic success wrapper, only DTOs.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Generic, TypeVar

from app.shared.web.errors import AppError

T = TypeVar("T")
R = TypeVar("R")


@dataclass(frozen=True, slots=True)
class Result(Generic[T]):
    _value: T | None
    _error: AppError | None
    is_success: bool

    @staticmethod
    def success(value: T) -> "Result[T]":
        return Result(value, None, True)

    @staticmethod
    def failure(error: AppError) -> "Result[T]":
        return Result(None, error, False)

    @property
    def value(self) -> T:
        if not self.is_success:
            raise ValueError("Result has no value — it is a failure.")
        return self._value  # type: ignore[return-value]

    @property
    def error(self) -> AppError:
        if self.is_success:
            raise ValueError("Result has no error — it is a success.")
        return self._error  # type: ignore[return-value]

    def match(self, on_success: Callable[[T], R], on_failure: Callable[[AppError], R]) -> R:
        return on_success(self._value) if self.is_success else on_failure(self._error)  # type: ignore[arg-type]
